src/worker.cc

```diff
@@ -0,0 +1,715 @@
+}
+
+void WorkerContext::Terminate(const FunctionCallbackInfo<Value>& args) {
+  WorkerContext* context = node::Unwrap<WorkerContext>(args.Holder());
+  if (context == nullptr)
+    return;
+  context->Terminate();
+}
+
+void WorkerContext::Terminate(bool should_emit_exit_event) {
+  CHECK_CALLED_FROM_OWNER(this);
+  ScopedLock::Mutex init_lock(initialization_mutex());
+  ScopedLock::Mutex api_lock(api_mutex());
+  ScopedLock::Mutex term_lock(termination_mutex());
+
+  if (termination_kind() != NONE)
+    return;
+
+  set_termination_kind(ABRUPT);
+
+  if (is_initialized()) {
+    should_emit_exit_event_ = should_emit_exit_event;
+    worker_isolate()->TerminateExecution();
+    worker_notifications()->Notify();
+  }
+}
+
+bool WorkerContext::ProcessMessageToWorker(Isolate* isolate,
+                                           WorkerMessage* message) {
+  Local<Value> argv[] = {
+    String::NewFromUtf8(isolate, message->payload()),
+    Integer::New(isolate, message->type())
+  };
+  delete message;
+  MakeCallback(isolate,
+               worker_wrapper(),
+               "_onmessage",
+               ARRAY_SIZE(argv),
+               argv);
+  return worker_env()->CanCallIntoJs();
+}
+
+void WorkerContext::ProcessMessagesToWorker() {
+  CHECK_CALLED_FROM_WORKER(this);
+
+  Isolate* isolate = worker_env()->isolate();
+  HandleScope scope(isolate);
+  Context::Scope context_scope(worker_env()->context());
+
+  flush_primary_queue:
+  uint32_t messages_processed = 0;
+  while (WorkerMessage* message = to_worker_messages_primary_.PopFront()) {
+    if (!ProcessMessageToWorker(isolate, message))
+      return;
+    messages_processed++;
+  }
+
+  if (messages_processed >= kMaxPrimaryQueueMessages) {
+    while (true) {
+      WorkerMessage* message = nullptr;
+      {
+        ScopedLock::Mutex lock(to_worker_messages_mutex());
+        message = to_worker_messages_backup_.PopFront();
+      }
+
+      if (message == nullptr)
+        goto flush_primary_queue;
+
+      if (!ProcessMessageToWorker(isolate, message))
+        return;
+    }
+  }
+}
+
+bool WorkerContext::ProcessMessageToOwner(Isolate* isolate,
+                                          WorkerMessage* message) {
+  Local<Value> argv[] = {
+    String::NewFromUtf8(isolate, message->payload()),
+    Integer::New(isolate, message->type())
+  };
+  delete message;
+  MakeCallback(isolate,
+               owner_wrapper(),
+               "_onmessage",
+               ARRAY_SIZE(argv),
+               argv);
+  return owner_env()->CanCallIntoJs();
+}
+
+void WorkerContext::ProcessMessagesToOwner() {
+  CHECK_CALLED_FROM_OWNER(this);
+  Isolate* isolate = owner_env()->isolate();
+  HandleScope scope(isolate);
+  Context::Scope context_scope(owner_env()->context());
+
+  flush_primary_queue:
+  uint32_t messages_processed = 0;
+  while (WorkerMessage* message = to_owner_messages_primary_.PopFront()) {
+    if (!ProcessMessageToOwner(isolate, message))
+      return;
```

Comment:
@petkaantonov If the condition above is satisfied and results in a `return` when the `to_owner_messages_primary_` ProducerConsumerQueue is full then the `to_owner_messages_backup_` queue will not be processed until `to_owner_messages_primary_` is once again full in the *next* call to to `WorkerContext::ProcessMessagesToOwner` - potentially much later or never. Is this as intended?

Also, queue event population/execution ordering is a concern in the pass off from the lock-free primary queue to locked backup queue and back again. If the last slot in the primary is refilled in the course of processing the primary PopFront while loop and the backup queue is subsequently populated and then in the course of the same primary PopFront while loop another slot becomes available and filled this seems to be a race condition.